lib/types.d.ts

```diff
@@ -0,0 +1,28 @@
+type MergeBy<TObject, TOtherObject> = Omit<TObject, keyof TOtherObject> & TOtherObject;
+
+type DeepRecord<TKey extends string | number | symbol, TValue> = {[key: string]: TValue | DeepRecord<TKey, TValue>};
+
+type TypeOptions = MergeBy<
+    {
+        keys: string;
+        collectionKeys: string;
+        values: Record<string, unknown>;
+    },
+    CustomTypeOptions
+>;
+
+interface CustomTypeOptions {}
+
+type Key = TypeOptions['keys'];
+type CollectionKey = TypeOptions['collectionKeys'];
+
+type OnyxKey = Key | CollectionKey;
+type KeyValueMapping = TypeOptions['values'];
+
+type Selector<TKey extends OnyxKey, TReturnType> = KeyValueMapping[TKey] extends object | string | number | boolean
```

Comment:
Let's remove string selector! 🚀 + I simplified the type a bit, is `KeyValueMapping[TKey] extends object | string | number | boolean` necessary? @fabioh8010 

```suggestion
type Selector<TKey extends OnyxKey, TReturnType> = KeyValueMapping[TKey] extends unknown
    ? ((value: unknown) => unknown)
    : ((value: KeyValueMapping[TKey] | null) => TReturnType);
```

Reply:
@blazejkustra We can't do in this way because anything is assignable to `unknown`, so every type will result in `(value: unknown) => unknown`.

[TS playground](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBAyhA2EDGwD2AnAPAFQHxQF4psoIAPYCAOwBMBnKAVyoGsrUB3KqAfigAoAbgEN4jCAC4mrdlwCUhfMzaduUoaPFTsCgvmwBuAFBHQkWAmRp0ARkIXEKDJjrB0ASyoBzfAHpfAiJiktIq8oqhslSm4NBwjtYATPbxVs7KUX4BGsFSGaq6SjKqQA)

Reply:
Oh, right! This should work: 
```ts
type Selector<T> = IsEqual<T, unknown> extends true ? (value: unknown) => unknown : (value: T) => T;
```

wdyt? @fabioh8010

Reply:
This one seems great @blazejkustra ! Tested here and it's working well so far.